feat(checkbox): add indeterminate checkbox support

DerekLouie opened the pull request:
@gmoothart @ErinCoughlan @jelbourn @ThomasBurleson 

This adds the indeterminate state to md-checkbox.

Please Review. 

References issue #1912


![indetcheckbox](https://cloud.githubusercontent.com/assets/709204/13895176/7069a010-ed2d-11e5-8dd5-09258b186094.png)
![indetunselectall](https://cloud.githubusercontent.com/assets/709204/13895177/7069fc9a-ed2d-11e5-931c-c65a1b762778.png)
![indetselectall](https://cloud.githubusercontent.com/assets/709204/13895175/706993ea-ed2d-11e5-8bec-7a2fb6695a84.png)

ThomasBurleson commented:
@DerekLouie - please note that we have a [PR and issue naming convention](https://github.com/angular/material/blob/master/CONTRIBUTING.md#-commit-message-format).

e.g. (for this PR)  `feat(checkbox): add indeterminate checkbox support`

DevVersion commented on src/components/checkbox/checkbox.js:
```diff
@@ -171,12 +197,48 @@ function MdCheckboxDirective(inputDirective, $mdAria, $mdConstant, $mdTheming, $
       }
+
+      function setIndeterminateState() {
+        var isIndeterminate = element.attr('indeterminate') === 'true' || false;
```
@DerekLouie I think it would be better, if you use 
```js
$mdUtils.parseAttributeValue(attr.indeterminate)
```

DerekLouie replied:
Ahh so the issue with this is that `$mdUtil.parseAttributeBoolean` will interpret an empty string as `true` which we want instead to be interpreted as `false` (in the case no value is passed in).

DevVersion replied:
@DerekLouie Yes, that's by intention. Shouldn't `<md-checkbox inderterminate>` work?

topherfangio replied:
The convention we usually follow is that if the attribute is found with no value, then it is treated as `true` as this follows standard HTML attributes (readonly, disabled, etc).

So,

 - `intermediate` == true
 - `intermediate="true"` == true
 - `intermediate="false"` == false

@EladBezalel @ThomasBurleson Please correct me if I'm wrong here.

DevVersion replied:
@topherfangio I agree with you. That's exactly what `$mdUtil.parseAttributeValue` does.

But notice, attributes like `readonly`, `disabled` or `required` should ignore false values.
That's why you should use in that case
```js
$mdUtil.parseAttributeValue(attr, false)
```

DerekLouie replied:
Ahh so it does seem like `parseAttributeValue` does not exist. BUT `parseAttributeBoolean` does exist. Unfortunately because we can accept a function now as well as a string, we can't use `parseAttributeBoolean` :(

DevVersion commented on src/components/checkbox/checkbox.js:
```diff
@@ -171,12 +197,48 @@ function MdCheckboxDirective(inputDirective, $mdAria, $mdConstant, $mdTheming, $
+          element[0].indeterminate = true;
+        } else {
+          element[0].indeterminate = undefined;
+        }
```
I think a ternary operator will be better here:

```js
element[0].indeterminate = isIndeterminate ? true : undefined;
```

DevVersion commented on src/components/checkbox/checkbox.scss:
```diff
@@ -105,6 +105,24 @@ $checkbox-top: 12px !default;
     cursor: default;
   }
 
+  &[indeterminate="true"] ._md-icon{
```
Can you not apply a custom class to the element (like `.md-indeterminate`)? 

This will be more compatible with the attribute value.

crisbeto commented on src/components/checkbox/checkbox.js:
```diff
@@ -69,6 +80,7 @@ function MdCheckboxDirective(inputDirective, $mdAria, $mdConstant, $mdTheming, $
 
   function compile (tElement, tAttrs) {
     var container = tElement.children();
+    var indeterminateStateEnabled = tElement[0].hasAttribute('indeterminate');
```
You can replace this one with `tAttrs.hasOwnProperty('indeterminate')` so you don't have to touch the DOM.

crisbeto commented on src/components/checkbox/demoSelectAll/script.js:
```diff
@@ -0,0 +1,41 @@
+    } else {
+      return false;
+    }
+  };
```
This whole check can be replaced with `return $scope.selected.length === $scope.items.length`.

crisbeto commented on src/components/checkbox/checkbox.js:
```diff
@@ -156,11 +175,18 @@ function MdCheckboxDirective(inputDirective, $mdAria, $mdConstant, $mdTheming, $
 
+        if (indeterminateStateEnabled) {
+          if (attr.indeterminateClick) {
+            scope.$apply(element.attr('indeterminate-click'));
```
You should be able to rewrite this one as `scope.$apply(attr.indeterminateClick);`.

DevVersion replied:
Also I think, that using `$apply` for that is not the right way.

Why aren't you using `$parse(attr.indeterminateClick)(scope)`?

crisbeto replied:
Isn't that the same as `scope.$eval(attr.indeterminateClick)`?

DevVersion replied:
:+1: Oh I think an `$evalAsync` is better here, because it's evaling in an listener. Thoughts?

crisbeto replied:
Looks like it. Can't be sure since I haven't used `$evalAsync` myself.

DerekLouie replied:
Yes, eval would be the correct thing to use here, and I should have noticed the `if` statement had the `attr` I needed :+1:.

As a sidenote: I tried substituting `$evalAsync` in here but it breaks necessary indet state logic.

This is because if this runs asynchronously it does not skip the code I need it to in the `render` function which ensures when a checkbox is indeterminate it does not adhere to the normal "checked' / 'unchecked' click logic.

topherfangio replied:
I ran into some issues with the Speed Dial when I was using `$apply` instead of `$evalAsync` because there were cases when the code could be called during an existing digest and throw an error.

If I recall, this occurred when a user added a custom `ng-click` to the component, so you may want to double-check that case.

DerekLouie replied:
Added an ng-click to the demo locally, both functions fired correctly, so we are all good on this front!

crisbeto commented on src/components/checkbox/checkbox.js:
```diff
@@ -156,11 +175,18 @@ function MdCheckboxDirective(inputDirective, $mdAria, $mdConstant, $mdTheming, $
       }
+
       function listener(ev) {
         if (element[0].hasAttribute('disabled')) {
```
This check can be rewritten as `attr.hasOwnProperty('disabled')`.

crisbeto commented on src/components/checkbox/checkbox.js:
```diff
@@ -171,12 +197,48 @@ function MdCheckboxDirective(inputDirective, $mdAria, $mdConstant, $mdTheming, $
+          } else {
+            element.removeClass(CHECKED_CSS);
+          }
+        }
```
This whole block could be shortened down to:

```javascript
var isChecked = scope.$eval(attr.indeterminateCheckedWhen);

element.toggleClass(CHECKED_CSS, isChecked);
isIndeterminate = !isChecked;
```

DerekLouie replied:
I do not believe this is logically equivalent

crisbeto replied:
Fair enough, I'm probably missing a bit of context.

crisbeto commented on src/components/checkbox/demoSelectAll/script.js:
```diff
@@ -0,0 +1,41 @@
+      return true;
+    } else {
+      return false;
+    }
```
This check can be replaced with `return $scope.selected.length !== 0 && $scope.selected.length !== $scope.items.length;`.

crisbeto commented on src/components/checkbox/demoSelectAll/script.js:
```diff
@@ -0,0 +1,41 @@
+  $scope.toggle = function (item, list) {
+    var idx = list.indexOf(item);
+    if (idx > -1) list.splice(idx, 1);
+    else list.push(item);
```
Not sure whether writing an `if/else` like this matches the guidelines.

DerekLouie replied:
Yeah I noticed that (this style got Apple in a lot of trouble not too long ago..), this code is pasted from [this file](https://github.com/angular/material/blob/master/src/components/checkbox/demoSyncing/script.js)

Should I can change it in both places?

DerekLouie commented:
Hello All,

I did a minor refactor. 

The appearance of the indeterminate state is now class based instead of attribute selector based. This allowed me to get rid of the "indeterminate-when" attribute and simplified the api a little.

I also fixed the aforementioned comments.

Please Review.

Best,

Derek

topherfangio commented on src/components/checkbox/checkbox.spec.js:
```diff
@@ -246,5 +247,115 @@ describe('mdCheckbox', function() {
+
+    describe('with the indeterminate attribute', function() {
+
+      it('should set indeterminate attr to false by default', function() {
```
I disagree with this test; I believe it should be set to true by default to follow our existing conventions.

DerekLouie replied:
Hmm.. that does make a lot of sense :).

Changed, thank you.

ThomasBurleson commented on src/components/checkbox/checkbox.js:
```diff
@@ -27,6 +27,16 @@ angular
+ * @param {expression=} indeterminate-checked-when This determines when the
+ * indeterminate checkbox should appear checked. The 'indeterminate' state is mutually
+ * exclusive with the 'checked' and 'unchecked' states.
+ * @param {expression=} indeterminate-click This gets triggered when the
```
I do not understand the feature here. At first glance, this seems to interfere with ng-click. Or is this an expression that should be invoked with the indeterminant state changes?

e.g.  `md-change-indeterminant`

What is the use case here ?

DerekLouie replied:
You are correct, this is not a good attribute.

Previously when I was unsure if a check box could be selected AND indeterminate I thought the click function would be useful for special toggle cases, but in retrospect this was a poor api design decision. Removing it now.

ThomasBurleson commented on src/components/checkbox/checkbox.js:
```diff
@@ -27,6 +27,16 @@ angular
+ * If a truthy expression or no value is passed in the checkbox renders in the indeterminate state.
+ * If falsy expression is passed in it just looks like a normal unchecked checkbox.
+ * To use the other indeterminate-* attributes this attribute must be present.
+ * @param {expression=} indeterminate-checked-when This determines when the
```
This appears to be a confusing API. Perhaps:

`md-indeterminant="<expression>"`

when the `expression` becomes true forces the checkbox to indeterminant state.

DerekLouie replied:
I have made some naming changes. Not completely sure if this is what you meant?

Please give the new code a look-over, any notes on how to give the api more clarity are appreciated!

ErinCoughlan commented on src/components/checkbox/checkbox.spec.js:
```diff
@@ -246,5 +247,96 @@ describe('mdCheckbox', function() {
+
+        pageScope.$apply();
+
+        expect(checkbox.hasClass(INDETERMINATE_CSS)).toBe(true);
```
Is `toHaveClass` copied over from our branch? This would be a lot easier to read as:

`expect(checkbox).toHaveClass(INDETERMINATE_CSS);`

ErinCoughlan commented on src/components/checkbox/checkbox.spec.js:
```diff
@@ -246,5 +247,96 @@ describe('mdCheckbox', function() {
+    describe('with the md-indeterminant attribute', function() {
+
+      it('should set md-indeterminant attr to true by default', function() {
+        var element = compileAndLink(
```
Do you need the wrapper `div`? Would `var checkbox = compileAndLink('<md-checkbox ... </md-checkbox>');` work?

ErinCoughlan commented on src/components/checkbox/checkbox.spec.js:
```diff
@@ -246,5 +247,96 @@ describe('mdCheckbox', function() {
+
+        var checkbox = element.find('md-checkbox');
+
+        pageScope.$apply();
```
`compileAndLink` calls `pageScope.$apply();`. Do you need another one?

ErinCoughlan commented on src/components/checkbox/demoBasicUsage/index.html:
```diff
@@ -43,6 +43,18 @@
+        </div>
+        <div flex-xs flex="50">
+          <md-checkbox md-indeterminant aria-label="Checkbox Disabled Indeterminate"
+            ng-disabled="true" class="md-primary">
```
nit: indent to line up with md-indeterminate or indent 4 spaces instead of 2 to separate attributes from the checkbox content. (Ditto above)

DerekLouie commented:
Hello All!

Did a significant re-write (thank you @ErinCoughlan for the help). 

Please check it out!

Best,

Derek

jelbourn commented on src/components/checkbox/checkbox.js:
```diff
@@ -27,7 +27,12 @@ angular
  * @param {string=} aria-label Adds label to checkbox for accessibility.
- * Defaults to checkbox's text. If no default text is found, a warning will be logged.
+ *     Defaults to checkbox's text. If no default text is found, a warning will be logged.
+ * @param {expression=} md-indeterminant This determines when the checkbox
```
Spelling: `indeterminate` here and everywhere else

jelbourn commented on src/components/checkbox/checkbox.js:
```diff
@@ -27,7 +27,12 @@ angular
+ *     should be rendered as 'indeterminant'.
+ *     If a truthy expression or no value is passed in the checkbox renders in the md-indeterminant state.
+ *     If falsy expression is passed in it just looks like a normal unchecked checkbox.
+ *     To use the other md-indeterminant-* attributes this attribute must be present.
```
Documentation needs to explain the interaction between `indeterminate` and `checked`

DevVersion replied:
In addition to that, I thought `md-indeterminat-*` attribute have been dropped? So maybe remove that line.

jelbourn commented on src/components/checkbox/checkbox.js:
```diff
@@ -89,8 +95,14 @@ function MdCheckboxDirective(inputDirective, $mdAria, $mdConstant, $mdTheming, $
       ngModelCtrl = ngModelCtrl || $mdUtil.fakeNgModel();
       $mdTheming(element);
+      if (mdIndeterminantStateEnabled) {
+        // Bootstrap initial state
```
This comment doesn't say anything useful.

jelbourn commented on src/components/checkbox/checkbox.js:
```diff
@@ -171,12 +184,18 @@ function MdCheckboxDirective(inputDirective, $mdAria, $mdConstant, $mdTheming, $
       }
+
+      function setIndeterminantState() {
+        isIndeterminant = scope.$eval(attr.mdIndeterminant) === false ? false : true;
```
`$scope.$watch` evaluates the expression already, so running `$eval` again is doing extra, unnecessary work.

jelbourn replied:
This assignment is kind of clunky. I'd do
```js
isIndeterminant = indeterminateValue || indeterminateValue === '';
```

DerekLouie replied:
Hmm so, the $scope.$watch will pass in the new value as the first parameter which will either be - `undefined`, `true`, or `false`.

If it returns `false` we want to set `isIndeterminate` to `false`. 

All other cases we want to set  `isIndeterminate` to `true`.

I cannot seem to find a better way to achieve that, but I am open to suggestions!

jelbourn commented on src/components/checkbox/checkbox.js:
```diff
@@ -171,12 +184,18 @@ function MdCheckboxDirective(inputDirective, $mdAria, $mdConstant, $mdTheming, $
+
+      function setIndeterminantState() {
+        isIndeterminant = scope.$eval(attr.mdIndeterminant) === false ? false : true;
+        element[0].indeterminate = isIndeterminant ? true : undefined;
```
Why is setting a property to the element necessary, and why does it need to be either `true` or `undefined`?

DerekLouie replied:
I believe this is to leverage some native browser implementation features. @ErinCoughlan?

DevVersion commented on src/components/checkbox/checkbox.spec.js:
```diff
@@ -246,5 +247,45 @@ describe('mdCheckbox', function() {
+        expect(checkbox).toHaveClass(INDETERMINATE_CSS);
+      });
+
+      it('should set its HTMLInputElement.indeterminate to be undefined when not md-indeterminant', function() {
```
I would rather say: `should not set the indeterminate property when md-indeterminate is false.`

jelbourn commented on src/components/checkbox/demoSelectAll/style.css:
```diff
@@ -0,0 +1,13 @@
+legend {
```
Prefix demo styles with `demo`-

jelbourn commented on src/core/style/mixins.scss:
```diff
@@ -251,4 +251,21 @@
 
+  &.md-indeterminant ._md-icon{
+    &:after {
+        box-sizing: border-box;
```
Too much indent. You're IDE is likely configured to the wrong indent for `.scss` files.

DevVersion commented on src/components/checkbox/checkbox.spec.js:
```diff
@@ -246,5 +247,45 @@ describe('mdCheckbox', function() {
+
+      it('should be set "md-indeterminant" class according to a passed in function', function() {
+        pageScope.isIndeterminate = function() { return true; } ;
+
```
@DerekLouie Maybe you can also add a test, which changes the `indeterminate` expression on runtime as well?

EladBezalel commented on src/core/style/mixins.scss:
```diff
@@ -251,4 +251,21 @@
     cursor: default;
   }
 
+  &.md-indeterminate ._md-icon{
```
Why is that in here and not in `checkbox.scss`?

ErinCoughlan replied:
The checkbox CSS is used both within the checkbox and the select menu (for multi-select case).

DevVersion replied:
@DerekLouie: Nitpicky -  Please space bracket from `._md-icon`

EladBezalel commented on src/core/style/themes.scss:
```diff
@@ -39,4 +39,10 @@ html, body {
     border-color: '{{primary-contrast-0.87}}';
   }
+
+  & .md-indeterminate[disabled] {
```
Why is that in here and not in `checkbox.scss`?

DerekLouie replied:
We created a mixin for the checkbox's css so it could be used in other components. This is where all the css goes now that might be used by multiple components.

EladBezalel replied:
Indeterminate isn't used anywhere else, only on progress and this seems not related, 
Also don't you think this class should be private?

topherfangio replied:
@EladBezalel In response to private: I had a discussion with Thomas where we thought that CSS classes representing the state of a component should probably be public as they are likely the ones to be used most.

DerekLouie commented:
Hello,

I have addressed the aforementioned comments, please review!

Best,

Derek

DevVersion commented on test/angular-material-spec.js:
```diff
@@ -194,10 +200,71 @@
+       *   {angular.JQLite} - The result of a jQuery query.
+       *   {Element} - A DOM element.
+       */
+      toContainHtml: function() {
```
Why has that been added?

ErinCoughlan replied:
I'm okay with waiting to add this method until we merge feat/panel. It was added here for completeness and convenience, but it's not being used yet.

DerekLouie replied:
I believe I these methods will be useful, and am adding all of them now (instead of just the `toHaveClass` method changes) so later down the line we don't get merge conflicts when we merge feat/panel.

DevVersion replied:
Didn't see actual point, because we rarely change the spec shims. So there won't be any merge conflicts. But it's okay.

jelbourn replied:
It's better to not add things until they're used. Plus you'll prevent merge conflicts.

DevVersion replied:
Yep, I agree. These test helpers, should be just merged within the `feat/panel` branch. Partially merging code snippets doesn't make any sense.

DerekLouie replied:
Fair point, taking out the non-used additions!

ErinCoughlan commented:
LGTM.

Make sure to change your commit message to be:
feat(checkbox): add indeterminate checkbox support

DerekLouie commented:
Changed the commit message, please review again at your earliest convenience!

jelbourn commented on src/components/checkbox/checkbox.js:
```diff
@@ -171,12 +183,18 @@ function MdCheckboxDirective(inputDirective, $mdAria, $mdConstant, $mdTheming, $
       }
+
+      function setIndeterminateState(newValue) {
+        isIndeterminate = (newValue === false) ? false : true;
```
isIndeterminate = newValue !== false;

jelbourn commented on src/components/checkbox/checkbox.js:
```diff
@@ -171,12 +183,18 @@ function MdCheckboxDirective(inputDirective, $mdAria, $mdConstant, $mdTheming, $
+
+      function setIndeterminateState(newValue) {
+        isIndeterminate = (newValue === false) ? false : true;
+        element[0].indeterminate = isIndeterminate ? true : undefined;
```
It's still not clear why you need to set an `indeterminate` property on the DOM element.

DerekLouie replied:
Apologies, I miss-understood the spec I was looking at. This is for when we move to the native checkbox. The only benefit of this would be that users can use the `:indeterminate` pseudo selector to find the indeterminate `md-checkbox`es.

@jelbourn I am okay with removing this code, do you care whether or not the directive is query-able using the  `:indeterminate` pseudo selector?

jelbourn commented on src/components/checkbox/checkbox.spec.js:
```diff
@@ -246,5 +247,64 @@ describe('mdCheckbox', function() {
+      it('should not set the HTMLInputElement.indeterminate property when md-indeterminate is false', function() {
+        var checkbox = compileAndLink('<md-checkbox md-indeterminate="false"></md-checkbox>');
+
+        expect(checkbox[0].indeterminate).toBe(undefined);
```
`toBeUndefined()`

jelbourn commented on src/components/checkbox/demoSelectAll/style.css:
```diff
@@ -0,0 +1,15 @@
+  .select-all-checkboxes {
+    padding-left: 30px;
+  }
+}
```
Just do these as
```css
.demo-legend { ... }
.demo-fieldset { ... }
.demo-select-all-checkboxes { .. }
```

ErinCoughlan commented:
@jelbourn Should there be aria-checked somewhere in here? I didn't see it in the initial code, so is it not helpful?

DerekLouie commented:
Took out some extra code and also added some code to dictate the 'aria-checked' attribute.

Please review.

ErinCoughlan commented on src/components/checkbox/checkbox.js:
```diff
@@ -171,12 +183,20 @@ function MdCheckboxDirective(inputDirective, $mdAria, $mdConstant, $mdTheming, $
+      function setIndeterminateState(newValue) {
+        isIndeterminate = newValue !== false;
+        if (isIndeterminate) {
+          element.attr('aria-checked', 'mixed');
```
aria-checked is never set for checked or unchecked checkboxes either. If you're going to set it to "mixed", you also should set it to true/false where appropriate.

DerekLouie replied:
It actually is set, and accounted for in the [tests](https://github.com/DerekLouie/material/blob/inteterminateCheckbox/src/components/checkbox/checkbox.spec.js#L43). It was a little curious though that the code isn't found in this file..

jelbourn replied:
I believe it's already handled by ngAria

DerekLouie commented:
Hello All,

I have addressed all the aforementioned comments, please review at your earliest convenience!

Best,

Derek

DevVersion commented:
LGTM now, aside one comment.

DerekLouie commented:
Apologies, there have been quite a few comments, I will happily address it, which comment?

DerekLouie commented:
Ahh thank you @DevVersion, fixed the comment.

jelbourn commented:
LGTM

EladBezalel commented on src/core/style/mixins.scss:
```diff
@@ -251,4 +251,21 @@
     cursor: default;
   }
 
+  &.md-indeterminate ._md-icon {
```
This is not related here as well, this is very specific for checkbox

DerekLouie replied:
@EladBezalel I could remove the `&`. The rule would still be valid and it would be one class less specific?

crisbeto commented:
LGTM

ThomasBurleson commented:
@DerekLouie - really nice feature. I LOVE this demo:

![screen shot 2016-03-30 at 12 27 27 pm](https://cloud.githubusercontent.com/assets/210413/14151277/ce7c3494-f672-11e5-8382-4824552249ad.png)
